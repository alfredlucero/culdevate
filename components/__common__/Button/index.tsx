import React from "react";
import styled, { css } from "styled-components";
import { culdevateDefaultTheme } from "../../defaultTheme";

interface StyledButtonProps {
  onClick: (event: React.MouseEvent) => void;
  disabled: boolean;
  kind: ButtonKind;
  type: ButtonType;
  size: ButtonSize;
}
const StyledButton = styled.button<StyledButtonProps>`
  border: none;
  border-radius: 2px;
  cursor: pointer;

  ${props =>
    props.size === "small" &&
    css`
      padding: 1rem;
      font-size: 1rem;
    `};

  ${props =>
    props.size === "medium" &&
    css`
      padding: 1.5rem;
      font-size: 1.5rem;
    `};

  ${props =>
    props.size === "large" &&
    css`
      padding: 2rem;
      font-size: 2rem;
    `};

  ${props =>
    props.kind === "primary" &&
    css`
      background-color: ${props.disabled ? props.theme.colors.gray : props.theme.colors.blue};
      color: ${props.theme.colors.white};
    `}
  
  ${props =>
    props.kind === "secondary-primary" &&
    css`
      border: ${props.disabled ? props.theme.colors.gray : props.theme.colors.blue} 2px solid;
      background-color: ${props.theme.colors.white};
      color: ${props.disabled ? props.theme.colors.gray : props.theme.colors.blue};
    `}

  ${props =>
    props.kind === "danger" &&
    css`
      background-color: ${props.disabled ? props.theme.colors.gray : props.theme.colors.red};
      color: ${props.theme.colors.white};
    `}

  ${props =>
    props.kind === "secondary-danger" &&
    css`
      border: ${props.disabled ? props.theme.colors.gray : props.theme.colors.red} 2px solid;
      background-color: ${props.theme.colors.white};
      color: ${props.disabled ? props.theme.colors.gray : props.theme.colors.red};
    `}
`;
StyledButton.defaultProps = {
  theme: culdevateDefaultTheme,
};

export type ButtonKind = "primary" | "danger" | "secondary-primary" | "secondary-danger";
export type ButtonType = "button" | "submit" | "reset";
export type ButtonSize = "small" | "medium" | "large";

interface ButtonProps {
  onClick: (event: React.MouseEvent) => void;
  children: React.ReactNode;
  kind: ButtonKind;
  type: ButtonType;
  size: ButtonSize;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  onClick,
  children,
  kind,
  type,
  size,
  disabled = false,
  ...passThroughProps
}) => {
  return (
    <StyledButton onClick={onClick} kind={kind} type={type} size={size} disabled={disabled} {...passThroughProps}>
      {children}
    </StyledButton>
  );
};

export default Button;
