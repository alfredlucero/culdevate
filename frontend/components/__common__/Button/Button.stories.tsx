import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import Button from "./index";

storiesOf("common/Button", module)
  .add("Primary Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="primary" size="small">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="primary" size="medium">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="primary" size="medium" disabled={true}>
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="primary" size="large">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="secondary" size="small">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondary" size="medium">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondary" size="medium" disabled={true}>
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="secondary" size="large">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Danger Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="danger" size="small">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="danger" size="medium">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="danger" size="medium" disabled={true}>
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="danger" size="large">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="secondaryDanger" size="small">
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondaryDanger" size="medium">
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondaryDanger" size="medium" disabled={true}>
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="secondaryDanger" size="large">
        Secondary Danger
      </Button>
    </StoryFrame>
  ));
