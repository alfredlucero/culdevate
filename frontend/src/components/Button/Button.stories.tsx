import React from "react";
import { storiesOf } from "@storybook/react";
import Button from "./index";

storiesOf("Common/Button", module).add("Default", () => (
  <Button variant="primary" onClick={() => {}}>
    Primary
  </Button>
));
