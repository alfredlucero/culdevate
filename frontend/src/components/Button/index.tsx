import React from "react";
import "./index.css";

interface ButtonProps {
  variant: ButtonVariant;
  children: React.ReactNode;
  onClick: (e: React.MouseEvent) => void;
}

type ButtonVariant = "primary" | "secondary" | "danger";

const Button: React.FC<ButtonProps> = ({ variant, children, onClick, ...passThroughProps }) => {
  return (
    <button className="btn btn-primary btn-primary:hover" {...passThroughProps} onClick={onClick}>
      {children}
    </button>
  );
};

export default Button;
