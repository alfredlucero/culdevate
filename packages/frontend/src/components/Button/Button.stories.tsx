import React from "react";
import { storiesOf } from "@storybook/react";
import Button from "./index";

storiesOf("Common/Button", module)
  .add("Primary", () => (
    <Button type="button" variant="primary" onClick={() => {}}>
      Primary
    </Button>
  ))
  .add("Primary Loading", () => (
    <Button type="button" variant="primary" loading={true} onClick={() => {}}>
      Primary Loading...
    </Button>
  ))
  .add("Primary Disabled", () => (
    <Button type="submit" variant="primary" onClick={() => {}} disabled={true}>
      Primary Disabled
    </Button>
  ))
  .add("Secondary", () => (
    <Button type="reset" variant="secondary" onClick={() => {}}>
      Secondary
    </Button>
  ))
  .add("Secondary Loading", () => (
    <Button type="reset" variant="secondary" loading={true} onClick={() => {}}>
      Secondary Loading...
    </Button>
  ))
  .add("Secondary Disabled", () => (
    <Button type="button" variant="secondary" onClick={() => {}} disabled={true}>
      Secondary Disabled
    </Button>
  ))
  .add("Danger", () => (
    <Button type="reset" variant="danger" onClick={() => {}}>
      Danger
    </Button>
  ))
  .add("Danger Loading", () => (
    <Button type="reset" variant="danger" loading={true} onClick={() => {}}>
      Danger Loading...
    </Button>
  ))
  .add("Danger Disabled", () => (
    <Button type="button" variant="danger" onClick={() => {}} disabled={true}>
      Danger Disabled
    </Button>
  ));
