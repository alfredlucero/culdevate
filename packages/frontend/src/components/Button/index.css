.btn {
  @apply font-bold py-2 px-4 rounded;
}

.btn-disabled {
  @apply opacity-50 cursor-not-allowed;
}

.btn-primary {
  @apply bg-blue-500 text-white;
}

.btn-primary:hover {
  @apply bg-blue-700;
}

.btn-secondary {
  @apply bg-transparent text-blue-700 border border-blue-500;
}

.btn-secondary:hover {
  @apply bg-blue-500 text-white border-transparent;
}
