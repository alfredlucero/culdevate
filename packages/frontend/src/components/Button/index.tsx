import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";
import "./index.css";

interface ButtonProps extends CommonProps {
  variant: ButtonVariant;
  type: ButtonType;
  children: React.ReactNode;
  onClick: (e: React.MouseEvent) => void;
  disabled?: boolean;
}

type ButtonType = "submit" | "button" | "reset";
type ButtonVariant = "primary" | "secondary" | "danger";

const buttonPrimaryClasses = ["btn-primary", "btn-primary:hover"];
const buttonSecondaryClasses = ["btn-secondary", "btn-secondary:hover"];
// TODO: implement these danger classes
const buttonDangerClasses = ["btn-danger", "btn-danger:hover"];

const Button: React.FC<ButtonProps> = ({
  variant,
  type,
  children,
  onClick,
  disabled = false,
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <button
      className={`${cn(
        "btn",
        ...(variant === "primary" ? buttonPrimaryClasses : []),
        ...(variant === "secondary" ? buttonSecondaryClasses : []),
        ...(variant === "danger" ? buttonDangerClasses : []),
        {
          "btn-disabled": disabled,
        },
      )} ${className}`}
      onClick={onClick}
      disabled={disabled}
      type={type}
      {...(testId !== "" ? { "data-testid": testId } : {})}
      {...passThroughProps}
    >
      {children}
    </button>
  );
};

export default Button;
