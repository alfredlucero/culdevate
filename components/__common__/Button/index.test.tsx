import React from "react";
import { render, cleanup } from "react-testing-library";
import Button from "./index";

afterEach(cleanup);

describe("Button", () => {
  test("should be true", () => {
    const { getByTestId } = render(<Button testId="testButton" />);

    expect(getByTestId("testButton")).toBeTruthy();
  });
});
