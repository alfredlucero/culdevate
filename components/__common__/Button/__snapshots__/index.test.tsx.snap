// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render Button component 1`] = `
<button
  class="sc-bdVaJa eCdspm"
  kind="primary"
  type="button"
>
  Primary
</button>
`;
